crypto/hpke/hpke.c

```diff
@@ -0,0 +1,2366 @@
+#include <openssl/rand.h>
+#include <openssl/kdf.h>
+#include <openssl/evp.h>
+#include <openssl/params.h>
+#include <openssl/param_build.h>
+#include <openssl/core_names.h>
+#include <internal/packet.h>
+#include <internal/common.h>
+#include <openssl/hpke.h>
+#include <internal/hpke_util.h>
+#include <openssl/err.h>
+
+/** default buffer size for keys and internal buffers we use */
+#ifndef OSSL_HPKE_MAXSIZE
+# define OSSL_HPKE_MAXSIZE 512
+#endif
+
+/* Define HPKE labels from RFC 9180 in hex for EBCDIC compatibility */
+/**< "HPKE" - "suite_id" label for 5.1 */
+static const char OSSL_HPKE_SEC51LABEL[] = "\x48\x50\x4b\x45";
+/**< "psk_id_hash" - in key_schedule_context */
+static const char OSSL_HPKE_PSKIDHASH_LABEL[] = "\x70\x73\x6b\x5f\x69\x64\x5f\x68\x61\x73\x68";
+/**<  "info_hash" - in key_schedule_context */
+static const char OSSL_HPKE_INFOHASH_LABEL[] = "\x69\x6e\x66\x6f\x5f\x68\x61\x73\x68";
+/**<  "base_nonce" - base nonce calc label */
+static const char OSSL_HPKE_NONCE_LABEL[] = "\x62\x61\x73\x65\x5f\x6e\x6f\x6e\x63\x65";
+/**<  "exp" - internal exporter secret generation label */
+static const char OSSL_HPKE_EXP_LABEL[] = "\x65\x78\x70";
+/**<  "sec" - external label for exporting secret */
+static const char OSSL_HPKE_EXP_SEC_LABEL[] = "\x73\x65\x63";
+/**<  "key" - label for use when generating key from shared secret */
+static const char OSSL_HPKE_KEY_LABEL[] = "\x6b\x65\x79";
+/**<  "psk_hash" - for hashing PSK */
+static const char OSSL_HPKE_PSK_HASH_LABEL[] = "\x70\x73\x6b\x5f\x68\x61\x73\x68";
+/**<  "secret" - for generating shared secret */
+static const char OSSL_HPKE_SECRET_LABEL[] = "\x73\x65\x63\x72\x65\x74";
+
+/*
+ * PEM header/footer for private keys
+ * PEM_STRING_PKCS8INF is just: "PRIVATE KEY"
+ */
+#define PEM_PRIVATEHEADER "-----BEGIN "PEM_STRING_PKCS8INF"-----\n"
+#define PEM_PRIVATEFOOTER "\n-----END "PEM_STRING_PKCS8INF"-----\n"
+
+/* max string len we'll try map to a suite */
+#define OSSL_HPKE_MAX_SUITESTR 38
+
+/* "strength" input to RAND_bytes_ex */
+#define OSSL_HPKE_RSTRENGTH 10
+
+/* an error macro just to make things easier */
+
+/*
+ * @brief info about an AEAD
+ */
+typedef struct {
+    uint16_t            aead_id; /**< code point for aead alg */
+    const char *        name;   /* alg name */
+    size_t              taglen; /**< aead tag len */
+    size_t              Nk; /**< size of a key for this aead */
+    size_t              Nn; /**< length of a nonce for this aead */
+} hpke_aead_info_t;
+
+/*
+ * @brief table of AEADs
+ */
+static hpke_aead_info_t hpke_aead_tab[] = {
+    { 0, NULL, 0, 0, 0 }, /* treat 0 as error so nothing here */
+    { OSSL_HPKE_AEAD_ID_AES_GCM_128, LN_aes_128_gcm, 16, 16, 12 },
+    { OSSL_HPKE_AEAD_ID_AES_GCM_256, LN_aes_256_gcm, 16, 32, 12 },
+#ifndef OPENSSL_NO_CHACHA20
+# ifndef OPENSSL_NO_POLY1305
+    { OSSL_HPKE_AEAD_ID_CHACHA_POLY1305, LN_chacha20_poly1305, 16, 32, 12 },
+# endif
+    { OSSL_HPKE_AEAD_ID_EXPORTONLY, LN_aes_128_gcm, 16, 16, 12 }
```

Comment:
These should probably be 0,0,0

Reply:
Could be. I need to add a test, so will revisit.

Reply:
So, did that, but in working it through I also changed my mind and went with what you (and richard B) earlier wanted for export-only stuff - there's now an ``OSSLHPKE_sender_export_encap()`` and similar for receiver that supports export only without requiring any plaintext/ciphertext - not sure it'll be used in real life but I guess it's better to have it and means that table entry is more as you'd expect